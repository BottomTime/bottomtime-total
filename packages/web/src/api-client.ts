import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
