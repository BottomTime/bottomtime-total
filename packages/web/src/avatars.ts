import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
