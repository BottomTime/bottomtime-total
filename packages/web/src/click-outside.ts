import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
