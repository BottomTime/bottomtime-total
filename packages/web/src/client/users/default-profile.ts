import { assertValid } from '@/helpers';
import { SuperAgentStatic } from 'superagent';
import {
  Profile,
  ProfileCertificationData,
  ProfileData,
  ProfileDataSchema,
  UserData,
} from './interfaces';
import { ProfileVisibility } from '@/constants';

export class DefaultProfile implements Profile {
  private static readonly JsonOmitKeys = new Set([
    'userId',
    'memberSince',
    'username',
  ]);

  constructor(
    private readonly agent: SuperAgentStatic,
    private readonly data: UserData,
  ) {}

  get userId(): string {
    return this.data.id;
  }
  get memberSince(): Date {
    return this.data.memberSince;
  }
  get username(): string {
    return this.data.username;
  }

  get avatar(): string | undefined {
    return this.data.profile.avatar;
  }
  set avatar(value: string | undefined) {
    this.data.profile.avatar = value;
  }

  get bio(): string | undefined {
    return this.data.profile.bio;
  }
  set bio(value: string | undefined) {
    this.data.profile.bio = value;
  }

  get birthdate(): string | undefined {
    return this.data.profile.birthdate;
  }
  set birthdate(value: string | undefined) {
    this.data.profile.birthdate = value;
  }

  get customData(): Record<string, unknown> | undefined {
    return this.data.profile.customData;
  }
  set customData(value: Record<string, unknown> | undefined) {
    this.data.profile.customData = value;
  }

  get certifications(): ProfileCertificationData[] | undefined {
    return this.data.profile.certifications;
  }
  set certifications(value: ProfileCertificationData[] | undefined) {
    this.data.profile.certifications = value;
  }

  get experienceLevel(): string | undefined {
    return this.data.profile.experienceLevel;
  }
  set experienceLevel(value: string | undefined) {
    this.data.profile.experienceLevel = value;
  }

  get location(): string | undefined {
    return this.data.profile.location;
  }
  set location(value: string | undefined) {
    this.data.profile.location = value;
  }

  get name(): string | undefined {
    return this.data.profile.name;
  }
  set name(value: string | undefined) {
    this.data.profile.name = value;
  }

  get profileVisibility(): ProfileVisibility {
    return this.data.profile.profileVisibility;
  }
  set profileVisibility(value: ProfileVisibility) {
    this.data.profile.profileVisibility = value;
  }

  get startedDiving(): string | undefined {
    return this.data.profile.startedDiving;
  }
  set startedDiving(value: string | undefined) {
    this.data.profile.startedDiving = value;
  }

  async save(): Promise<void> {
    const profile = assertValid<ProfileData>(
      this.data.profile,
      ProfileDataSchema,
    );
    await this.agent.put(`/api/profiles/${this.username}`).send(profile);
    this.data.profile = profile;
  }

  toJSON(): ProfileData {
    return assertValid(this.data.profile, ProfileDataSchema);
  }
}
