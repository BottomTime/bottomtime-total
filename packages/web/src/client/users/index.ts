export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
