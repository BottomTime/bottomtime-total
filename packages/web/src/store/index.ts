export * from './alerts.store';
export * from './current-user.store';
export * from './operators.store';
export * from './dive-sites.store';
export * from './error.store';
export * from './friends.store';
export * from './log-entries.store';
export * from './profiles.store';
export * from './tanks.store';
export * from './toasts.store';
