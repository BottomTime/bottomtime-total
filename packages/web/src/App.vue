<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
