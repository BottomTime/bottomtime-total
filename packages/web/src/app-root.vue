<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
