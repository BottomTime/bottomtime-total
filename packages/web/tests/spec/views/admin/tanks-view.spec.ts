import {
  ApiClient,
  Fetcher,
  ListTanksResponseDTO,
  ListTanksResponseSchema,
  Tank,
  TankDTO,
  TankMaterial,
  TankSchema,
} from '@bottomtime/api';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';
import { Router } from 'vue-router';

import { ApiClientKey } from '../../../../src/api-client';
import TanksListItem from '../../../../src/components/tanks/tanks-list-item.vue';
import { useCurrentUser } from '../../../../src/store';
import AdminTanksView from '../../../../src/views/admin/tanks-view.vue';
import { createRouter } from '../../../fixtures/create-router';
import TestTankData from '../../../fixtures/tanks.json';
import { AdminUser, BasicUser } from '../../../fixtures/users';

const AddButton = 'button#tanks-list-add';
const MaterialRadio = {
  Aluminum: '#material-al',
  Steel: '#material-fe',
} as const;
const NameInput = '#name';
const VolumeInput = '#volume';
const PressureInput = '#pressure';

const SaveButton = 'button#save-tank';

describe('Admin Tanks View', () => {
  let fetcher: Fetcher;
  let client: ApiClient;
  let router: Router;
  let tankData: ListTanksResponseDTO;

  let pinia: Pinia;
  let currentUser: ReturnType<typeof useCurrentUser>;
  let opts: ComponentMountingOptions<typeof AdminTanksView>;
  let listSpy: jest.SpyInstance;

  beforeAll(() => {
    fetcher = new Fetcher();
    client = new ApiClient({ fetcher });
    router = createRouter();
  });

  beforeEach(() => {
    tankData = ListTanksResponseSchema.parse(TestTankData);

    pinia = createPinia();
    currentUser = useCurrentUser(pinia);
    currentUser.user = AdminUser;

    opts = {
      global: {
        plugins: [pinia, router],
        provide: {
          [ApiClientKey as symbol]: client,
        },
        stubs: {
          teleport: true,
        },
      },
    };

    listSpy = jest.spyOn(client.tanks, 'listTanks').mockResolvedValue({
      tanks: tankData.tanks.map((tank) => new Tank(fetcher, tank)),
      totalCount: tankData.totalCount,
    });
  });

  it('will render the tanks list', async () => {
    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();
    expect(
      wrapper.get('[data-testid="tanks-list-counts"]').text(),
    ).toMatchSnapshot();

    const items = wrapper.findAllComponents(TanksListItem);
    expect(items).toHaveLength(tankData.tanks.length);
    expect(listSpy).toHaveBeenCalledWith();
  });

  it('will render login form if user is not authenticated', async () => {
    currentUser.user = null;

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    expect(wrapper.get('[data-testid="login-form"]').isVisible()).toBe(true);
    expect(wrapper.find('[data-testid="tanks-list-counts"]').exists()).toBe(
      false,
    );
    expect(listSpy).not.toHaveBeenCalled();
  });

  it('will render forbidden message if user is not an admin', async () => {
    currentUser.user = BasicUser;

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    expect(wrapper.get('[data-testid="forbidden-message"]').isVisible()).toBe(
      true,
    );
    expect(wrapper.find('[data-testid="tanks-list-counts"]').exists()).toBe(
      false,
    );
    expect(listSpy).not.toHaveBeenCalled();
  });

  it('will allow a user to delete a tank', async () => {
    const dto = tankData.tanks[3];
    const tank = new Tank(fetcher, dto);
    const spy = jest.spyOn(tank, 'delete').mockResolvedValue();
    jest.spyOn(client.tanks, 'wrapDTO').mockReturnValue(tank);
    jest.spyOn(client.tanks, 'listTanks').mockResolvedValue({
      tanks: [tank],
      totalCount: 1,
    });

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    await wrapper.get(`[data-testid="delete-tank-${dto.id}"]`).trigger('click');
    await wrapper.get('[data-testid="dialog-confirm-button"]').trigger('click');
    await flushPromises();

    expect(spy).toHaveBeenCalled();
    expect(wrapper.find(`[data-testid="select-tank-${dto.id}"]`).exists()).toBe(
      false,
    );
    expect(wrapper.find('[data-testid="dialog-confirm-button"]').exists()).toBe(
      false,
    );
  });

  it('will allow a user to change their mind about deleting a tank', async () => {
    const dto = tankData.tanks[5];
    const tank = new Tank(fetcher, dto);
    const spy = jest.spyOn(tank, 'delete').mockResolvedValue();
    jest.spyOn(client.tanks, 'wrapDTO').mockReturnValue(tank);

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    await wrapper.get(`[data-testid="delete-tank-${dto.id}"]`).trigger('click');
    await wrapper.get('[data-testid="dialog-cancel-button"]').trigger('click');
    await flushPromises();

    expect(spy).not.toHaveBeenCalled();
    expect(
      wrapper.find(`[data-testid="select-tank-${dto.id}"]`).isVisible(),
    ).toBe(true);
    expect(wrapper.find('[data-testid="dialog-cancel-button"]').exists()).toBe(
      false,
    );
  });

  it('will allow a user to create a new profile', async () => {
    const data: TankDTO = {
      id: 'e19dcf78-b7b5-4e12-8dbe-96a1ee498091',
      name: 'New Tank',
      volume: 12.5,
      workingPressure: 200,
      isSystem: true,
      material: TankMaterial.Aluminum,
    };

    const spy = jest
      .spyOn(client.tanks, 'createTank')
      .mockResolvedValue(new Tank(fetcher, data));

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    await wrapper.get(AddButton).trigger('click');
    await wrapper.get(NameInput).setValue(data.name);
    await wrapper.get(VolumeInput).setValue(data.volume);
    await wrapper.get(PressureInput).setValue(data.workingPressure);
    await wrapper.get(MaterialRadio.Aluminum).setValue(true);
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.find(NameInput).exists()).toBe(false);
    expect(spy).toHaveBeenCalledWith({
      ...data,
      id: '',
    });
    const item = wrapper.findComponent(TanksListItem);
    expect(item.text()).toContain(data.name);
  });

  it('will allow a user to edit a profile', async () => {
    const name = 'New Name';
    const volume = 12.5;
    const workingPressure = 200;
    const dto = tankData.tanks[8];
    const expected: TankDTO = {
      ...dto,
      name,
      volume,
      workingPressure,
      material: TankMaterial.Steel,
    };
    const tank = new Tank(fetcher, expected);
    const spy = jest.spyOn(tank, 'save').mockResolvedValue();
    let savedDTO: TankDTO | undefined;

    jest.spyOn(client.tanks, 'wrapDTO').mockImplementation((dto) => {
      savedDTO = TankSchema.parse(dto);
      return tank;
    });

    const tankSelector = `[data-testid="select-tank-${dto.id}"]`;

    const wrapper = mount(AdminTanksView, opts);
    await flushPromises();

    await wrapper.get(tankSelector).trigger('click');
    await wrapper.get(NameInput).setValue(name);
    await wrapper.get(VolumeInput).setValue(volume);
    await wrapper.get(PressureInput).setValue(workingPressure);
    await wrapper.get(MaterialRadio.Steel).setValue(true);
    await wrapper.get(SaveButton).trigger('click');
    await flushPromises();

    expect(wrapper.find(NameInput).exists()).toBe(false);
    expect(spy).toHaveBeenCalled();
    expect(savedDTO).toEqual(expected);

    expect(wrapper.find(tankSelector).text()).toBe(name);
  });
});
