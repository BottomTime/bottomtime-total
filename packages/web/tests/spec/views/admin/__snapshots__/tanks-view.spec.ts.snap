// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin Tanks View will render the tanks list 1`] = `"Showing 22 tank profile(s)"`;
